src/celeritas/ext/detail/HitProcessor.cc

```diff
@@ -380,14 +409,23 @@ void HitProcessor::operator()(DetectorStepOutput const& out, size_type i) const
  *
  * This is a bit like \c G4Step::UpdateTrack .
  */
-void HitProcessor::update_track(ParticleId id) const
+void HitProcessor::update_track(DetectorStepOutput const& out, size_type i) const
 {
+    ParticleId id = out.particle[i];
     CELER_EXPECT(id < tracks_.size());
     G4Track& track = *tracks_[id.unchecked_get()];
     step_->SetTrack(&track);
 
     G4ParticleDefinition const& pd = *track.GetParticleDefinition();
 
+    if (!out.primary_id.empty())
+    {
+        PrimaryId celeritas_primary_id = out.primary_id[i];
+        CELER_EXPECT(celeritas_primary_id < celeritas_to_g4_track_id_.size());
```

Comment:
```suggestion
        CELER_ASSERT(celeritas_primary_id < celeritas_to_g4_track_id_.size());
```
for internal consistency checks